### PR TITLE
#95 A&E Extract Processing

### PR DESCRIPTION
**Pull request by catherineholland1**
#95 

conversion of scripts:
B05a Process A&E (AE2) extract
B05b Link CUP marker to A&E
B05X A&E_tests

One of the factor labels `ae_reasonwait` does not match the current levels of the factor / way to update this? code is commented out.

Tried to look at creating a function for the testing as Outpatient and A&E tests very similar. However, they have one variable different in the flags. If this is centralised then some variables in on extract may be missed. To overcome this I have just created a function to match and combine the `slf_new` and `slf_existing`, compute the `difference` and `pct_change` and plot these where issues are flagged. This step should be the same for all so the function could work well, thoughts on this appreciated as this is just based off looking at testing for 2 of the `B` scripts.

### CODE REVIEW

**Comment by catherineholland1** [comment]
15/3/22 edit to the script

`create_ae_extract_flags.R` was created to produce the flags for the ae extract, this draws on functions already written `create_demog_test_flags.R` and `create_hb_test_flags.R`, in addition to new one `create_hb_costs_test_flags.R`. I had to edit the `demog_test_flags` to allow for postcode to not be included as it is not in the episode file compared. Hope this makes sense and is okay?

`produce_ae_extract_test.R` creates the summary of the flags. I tried to create specific functions that computed the summary for monthly costs, hb and hb costs all sep but there was an issue of bringing the summarise together hence why this is all done in the one function. Perhaps someone else can see a way of doing it? It does appear that if parts could be written then combined would be useful for other scripts but as `summarise` removes all other rows was proving problematic. This also has a postcode T/F input.

The same functions above then can be applied to the episode file. 

`produce_extract_comparison_test.R` combines the `slf_new` and `slf_existing` and computes the `difference` and pct_change` this should be able to be applied to all scripts. This then can be plotted (in each script) to show the differences.


Hope this all makes sense and runs okay!

**Comment by catherineholland1** [comment]
removed the dates from the summary - they seem to be throwing off the `pivot_longer` as all variables need to be the same type. Any ideas on this, or crucial to keep in?

File: Production_scripts/B05a_Process_A&E_(AE2)_extract.R [hunk 1]
```diff
@@ -0,0 +1,1094 @@
...
+
+
+ae_episode_extract <- readr::read_csv(
+  paste0(
+    get_year_dir(year = latest_year),
+    "/Extracts/A&E-episode-level-extract-20",
+    latest_year,
+    ".csv.gz"
+  )
+) %>%
```
Comment (Jennit07) [inline Production_scripts/B05a_Process_A&E_(AE2)_extract.R]: I think you mean get_boxi_extract which has now been pushed. We are tidying up PR's so i'll make this change :)

File: Production_scripts/B05a_Process_A&E_(AE2)_extract.R [hunk 2]
```diff
@@ -0,0 +1,1094 @@
...
+    cost_total_net = "Total Net Costs",
+    location = "Treatment Location Code",
+    case_ref_number = "Case Reference Number"
+  ) %>%
+  # date types
+  mutate(
+    dob = as.Date(dob),
+    record_keydate1 = as.Date(record_keydate1),
+    record_keydate2 = as.Date(record_keydate2)
+  )
```
Comment (Jennit07) [inline Production_scripts/B05a_Process_A&E_(AE2)_extract.R]: I've added in the boxi_extract_path function and while doing this i've added in using `col_type` which allows you to specify the formats. I've removed this

File: Production_scripts/B05a_Process_A&E_(AE2)_extract.R [hunk 3]
```diff
@@ -0,0 +1,1094 @@
...
+    #                         "Wait for treatment - Refers to the initial A&E treatment",
+    #                         "Wait for treatment - To commence",
+    #                         "Wait for treatment - To be completed",
+    #                         "Wait for diagnostic test(s)",
+    #                         "Wait for diagnostic test(s) - To be performed",
+    #                         "Awaiting results. Refers to results which will determine the next step in the patient journey",
+    #                         "Wait for first assessment",
+    #                         "Clinical reason(s)",
+    #                         "Other reason"
+    #                       )),
```
Comment (Jennit07) [inline Production_scripts/B05a_Process_A&E_(AE2)_extract.R:289]: Not sure why this is all commented out? Do we need this?

File: Production_scripts/B05a_Process_A&E_(AE2)_extract.R [hunk 4]
```diff
@@ -0,0 +1,1094 @@
...
+readr::write_rds(outfile,
+  paste0(
+    get_year_dir(year = latest_year),
+    "/a&e_data-20",
+    latest_year, ".zsav"
+  ),
+  compress = "gz"
+)
+
+
```
Comment (Jennit07) [inline Production_scripts/B05a_Process_A&E_(AE2)_extract.R]: use new function `get_source_extract_path` for saving out - changes made

File: Production_scripts/B05a_Process_A&E_(AE2)_extract.R [hunk 5]
```diff
@@ -0,0 +1,1094 @@
...
+readr::write_rds(outfile,
+  paste0(
+    get_year_dir(year = latest_year),
+    "/a&e_for_source-20",
+    latest_year, ".zsav"
+  ),
+  compress = "gz"
+)
+
+
```
Comment (Jennit07) [inline Production_scripts/B05a_Process_A&E_(AE2)_extract.R:934]: I've split this out into a new script but need to go back and have a look through

**Review by Jennit07 — CHANGES_REQUESTED** [review]
Made some changes and commented on the review where i've made the changes. Still need to go back to check the CUP marker script

File: Production_scripts/B05b_Link_CUP_marker.R [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+## get data ##
+ae_cup_extract <- readr::read_csv(
+  file =
+    paste0(
+      get_year_dir(year = latest_year),
+      "/Extracts/A&E-UCD-CUP-extract-20",
+      latest_year,
+      ".csv.gz"
+    )
```
Comment (Jennit07) [inline Production_scripts/B05b_Link_CUP_marker.R:32]: Use `get_boxi_extract` function here with col types - changes made

File: Production_scripts/B05b_Link_CUP_marker.R [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    keyTime1 = "ED Arrival Time",
+    case_ref_number = "ED Case Reference Number [C]",
+    cup_marker = "CUP Marker",
+    cup_pathway = "CUP Pathway Name"
+  ) %>%
+  # date type
+  mutate(
+    record_keydate1 = as.Date(record_keydate1)
+  )
+
```
Comment (Jennit07) [inline Production_scripts/B05b_Link_CUP_marker.R:46]: Can remove this now that `col_types` specifies the type

File: Production_scripts/B05b_Link_CUP_marker.R [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+library(slfhelper)
+library(readr)
+
+# Read in data---------------------------------------
+
+# Specify year
+year <- 1920
+
+## get data ##
+ae_cup_extract <- readr::read_csv(
```
Comment (Jennit07) [inline Production_scripts/B05b_Link_CUP_marker.R:25]:
```suggestion
ae_cup_file<- readr::read_csv(
```
Ive renamed this for reading the raw data

File: Production_scripts/B05b_Link_CUP_marker.R [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    case_ref_number = "ED Case Reference Number [C]",
+    cup_marker = "CUP Marker",
+    cup_pathway = "CUP Pathway Name"
+  ) %>%
+  # date type
+  mutate(
+    record_keydate1 = as.Date(record_keydate1)
+  )
+
+
```
Comment (Jennit07) [inline Production_scripts/B05b_Link_CUP_marker.R:47]: New section for data cleaning

File: Production_scripts/B05b_Link_CUP_marker.R [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+## sort for linking onto data extract ##
+# remove any duplicates
+ae_cup_extract <-
```
Comment (Jennit07) [inline Production_scripts/B05b_Link_CUP_marker.R:50]: Renamed this for cleaning section
```suggestion
ae_cup_clean <-
```

File: Production_scripts/B05b_Link_CUP_marker.R [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+  ) %>%
+  ungroup()
+
+
+## match files ##
+matched_data <-
+  outfile %>%
+  full_join(ae_cup_extract, by = c("record_keydate1", "keyTime1", "case_ref_number")) %>%
+  arrange(chi, record_keydate1, keyTime1, record_keydate2, keyTime2)
+
```
Comment (Jennit07) [inline Production_scripts/B05b_Link_CUP_marker.R:66]: Made a new section for joining the data. Read in the processed  A&E extract here

File: Production_scripts/B05b_Link_CUP_marker.R [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+# .rds file
+readr::write_rds(outfile,
+                 paste0(
+                   get_year_dir(year = latest_year),
+                   "/a&e_for_source-20",
+                   latest_year, ".zsav"
+                 ),
+                 compress = "gz"
+)
+
```
Comment (Jennit07) [inline Production_scripts/B05b_Link_CUP_marker.R:143]:
```suggestion
# Save as zsav file
outfile %>%
  readr::write_rds(get_source_extract_path(year, "AE", ext = "zsav"))

# Save as rds file
outfile %>%
  readr::write_rds(get_source_extract_path(year, "AE", ext = "rds"))
```
Changed to using `get_source_extract_path` for saving processed file

**Review by Jennit07 — CHANGES_REQUESTED** [review]

**Comment by Jennit07** [comment]
I've reverted the `get_demog_test_flags` function back to how it was because it was throwing up issues during checks. This function should already account for if postcode is missing or blank. In the A&E extract we do have the postcode variable as well, so I'm not sure what issue you had with it. If this persists we can create a new branch and alter the function if needed

**Comment by Jennit07** [comment]
I've split out your test functions which you mentioned above: 
- `create_ae_extract_flags`
- `produce_ae_extract_test`
- `produce_extract_comparison_test.R`
and i've put them onto an ae_tests branch. I'll have a look at these separately as i'm still trying to figure out the tests and once we have a format these can be applied to all the B file extracts

**Comment by Jennit07** [comment]
@Moohan I think i'm happy with this now to pass to you for review. I spent some time splitting tests and the link CUP marker out of the script. There was a few things the build check was not passing so i also spent some time fixing this. All checks now pass and i think these scripts are ready to merge (I had to change some of the functions mentioned above). I've applied the suggestions made to the script and just commented on review for where these were made (mainly so Catherine can look back and see suggestions). Feel free to agree/disagree with anything mentioned. All conversations should be able to be resolved as i have commented throughout/made the changes. Hope this is okay, thanks :)

**Review by Jennit07 — APPROVED** [review]

File: Production_scripts/B05a_Process_A&E_(AE2)_extract.R [hunk 13]
```diff
@@ -0,0 +1,950 @@
...
+library(readr)
+
+
+# Read in data---------------------------------------
+
+# Specify year
+year <- 1920
+
+ae_file <- readr::read_csv(
+  file = get_boxi_extract_path(year, "AE"), n_max = 2000,
```
Comment (Moohan) [inline Production_scripts/B05a_Process_A&E_(AE2)_extract.R:24]:
```suggestion
  file = get_boxi_extract_path(year, "AE"),
```

File: Production_scripts/B05a_Process_A&E_(AE2)_extract.R [hunk 14]
```diff
@@ -0,0 +1,950 @@
...
+    jul_cost = if_else(month == "07", cost_total_net, 0),
+    aug_cost = if_else(month == "08", cost_total_net, 0),
+    sep_cost = if_else(month == "09", cost_total_net, 0),
+    oct_cost = if_else(month == "10", cost_total_net, 0),
+    nov_cost = if_else(month == "11", cost_total_net, 0),
+    dec_cost = if_else(month == "12", cost_total_net, 0),
+    jan_cost = if_else(month == "01", cost_total_net, 0),
+    feb_cost = if_else(month == "02", cost_total_net, 0),
+    mar_cost = if_else(month == "03", cost_total_net, 0)
+  ) %>%
```
Comment (Moohan) [inline Production_scripts/B05a_Process_A&E_(AE2)_extract.R:149]: Could you create a function out of this code, please?

`allocate_single_day_costs <- function(data, date, cost_var)`

File: Production_scripts/B05a_Process_A&E_(AE2)_extract.R [hunk 15]
```diff
@@ -0,0 +1,950 @@
...
+  # sort for linking on CUP marker
+  arrange(record_keydate1, keyTime1, case_ref_number)
```
Comment (Moohan) [inline Production_scripts/B05a_Process_A&E_(AE2)_extract.R:151]: Don't need all these sorts in R. They're only required in SPSS for matching files together.

File: Production_scripts/B05a_Process_A&E_(AE2)_extract.R [hunk 16]
```diff
@@ -0,0 +1,950 @@
...
+        "Ambulance and A&E retrieval team. Includes both road and air ambulance modes of transport.",
+        "Out of Hours transport OOH has arranged transport (includes PTS transport but not emergency ambulance).",
+        "Private transport Includes car, taxi, motorbike, bicycle, etc",
+        "Public transport Includes bus, train, etc.",
+        "Walking On foot.",
+        "Police/prison transport Patient is brought to A&E in a police or prison vehicle.",
+        "Other The true value to be recorded is not covered by any of the specific given categories. Includes mortuary van.",
+        "Not known"
+      )
+    ),
```
Comment (Moohan) [inline Production_scripts/B05a_Process_A&E_(AE2)_extract.R:173]: Creating factors with a different label when the data is a character doesn't work / could break people's code. e.g.

```r
ae_clean %>% filter(ae_arrivalmode == "01")
```
Would usually work but won't after it's been turned into a factor. R drops the underlying data and replaces it with a number 1:n for the number of levels. 

We can discuss 😄

File: Production_scripts/B05a_Process_A&E_(AE2)_extract.R [hunk 17]
```diff
@@ -0,0 +1,950 @@
...
+    case_ref_number
+  )
+
+# Save a temp file for matching (.zsav)
+haven::write_sav(outfile,
+  paste0(
+    get_year_dir(year = year),
+    "/a&e_data-20",
+    latest_year, ".zsav"
+  ),
```
Comment (Moohan) [inline Production_scripts/B05a_Process_A&E_(AE2)_extract.R:936]: Use function for file path

**Review by Moohan — CHANGES_REQUESTED** [review]

File: Production_scripts/B05a_Process_A&E_(AE2)_extract.R (hunk 16 above)
Comment (catherineholland1) [reply Production_scripts/B05a_Process_A&E_(AE2)_extract.R:173]: @Moohan so remove the labels code? and keep numeric levels?

**Review by catherineholland1 — COMMENTED** [review]
updated/edits checked

File: Production_scripts/Process_extracts/extract_a&e.R [hunk 18]
```diff
@@ -0,0 +1,620 @@
...
+    cup_pathway = first(cup_pathway)
+  ) %>%
+  ungroup()
+
+
+# Join data--------------------------------------------
+
+matched_ae_data <- outfile %>%
+  full_join(ae_cup_clean, by = c("record_keydate1", "keyTime1", "case_ref_number")) %>%
+  arrange(chi, record_keydate1, keyTime1, record_keydate2, keyTime2)
```
Comment (Moohan) [inline Production_scripts/Process_extracts/extract_a&e.R:554]: I think this should just be a `left_join`. The records which have CUP marker should all have a match in the main data but if they don't we don't want to include random bits of data not linked to anything.

```suggestion
matched_ae_data <- outfile %>%
  left_join(ae_cup_clean, by = c("record_keydate1", "keyTime1", "case_ref_number")) %>%
  arrange(chi, record_keydate1, keyTime1, record_keydate2, keyTime2)
```

**Review by Moohan — CHANGES_REQUESTED** [review]
Just a few more small changes. Only other thing is the bits of test files on this branch which might cause conflicts later.

**Comment by catherineholland1** [comment]
> Only other thing is the bits of test files on this branch which might cause conflicts later.

Shall I delete the test files parts/scripts? as that was done before Jen started working on the tests

**Comment by Moohan** [comment]
> > Only other thing is the bits of test files on this branch which might cause conflicts later.
> 
> Shall I delete the test files parts/scripts? as that was done before Jen started working on the tests

Yep, I think that's probably simplest for now.

**Comment by catherineholland1** [comment]
test files deleted as not required - is this ready to be merged or another check required?

**Review by Moohan — APPROVED** [review]
I think all good now and finally ready to merge!

**Pull request merged**